File: dotnet-service/src/Internal/Infrastructure/Queue/WriteBehindWorker.cs
using System.Diagnostics;
using System.Text.Json;
using System.Text.Json.Serialization;
using DotnetService.Internal.Config;
using DotnetService.Internal.Domain.Events;
using DotnetService.Internal.Domain.Interfaces;
using DotnetService.Internal.Observability;
using Microsoft.Extensions.Hosting;
using StackExchange.Redis;

namespace DotnetService.Internal.Infrastructure.Queue;

public sealed class WriteBehindWorker : BackgroundService
{
    private readonly IConnectionMultiplexer _connection;
    private readonly string _streamName;
    private readonly string _groupName;
    private readonly string _consumerName;
    private readonly int _batchSize;
    private readonly TimeSpan _blockTimeout;
    private readonly IProductStore _store;
    private readonly ILoggerAdapter _logger;
    private readonly JsonSerializerOptions _jsonOptions;
    private readonly string _source;

    public WriteBehindWorker(
        IConnectionMultiplexer connection,
        AppConfig config,
        IProductStore store,
        ILoggerAdapter logger)
    {
        _connection = connection;
        _streamName = config.WriteBehind.StreamName;
        _groupName = "dotnet-service";
        _consumerName = $"{_groupName}-{Environment.ProcessId}-{Guid.NewGuid():N}";
        _batchSize = config.WriteBehind.BatchSize;
        _blockTimeout = config.WriteBehind.FlushInterval;
        _store = store;
        _logger = logger;
        _source = _groupName;
        _jsonOptions = new JsonSerializerOptions(JsonSerializerDefaults.Web);
        _jsonOptions.Converters.Add(new JsonStringEnumConverter(JsonNamingPolicy.CamelCase));
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        if (!_connection.IsConnected)
        {
            return;
        }
        await EnsureGroupAsync(stoppingToken).ConfigureAwait(false);
        var db = _connection.GetDatabase();
        while (!stoppingToken.IsCancellationRequested)
        {
            try
            {
                var entries = await db.StreamReadGroupAsync(
                    _streamName,
                    _groupName,
                    _consumerName,
                    ">",
                    count: _batchSize).ConfigureAwait(false);
                if (entries.Length == 0)
                {
                    await Task.Delay(_blockTimeout, stoppingToken).ConfigureAwait(false);
                    continue;
                }
                var batchTimer = Stopwatch.StartNew();
                var nowMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
                var maxLagSeconds = 0.0;
                foreach (var entry in entries)
                {
                    var lagSeconds = CalculateLagSeconds(nowMs, entry.Id);
                    if (lagSeconds > maxLagSeconds)
                    {
                        maxLagSeconds = lagSeconds;
                    }
                    await ProcessEntryAsync(db, entry, stoppingToken).ConfigureAwait(false);
                }
                batchTimer.Stop();
                long queueLength = -1;
                try
                {
                    queueLength = await db.StreamLengthAsync(_streamName).ConfigureAwait(false);
                }
                catch (Exception ex)
                {
                    _logger.Warn("failed to read write-behind queue length", new { Error = ex.Message });
                    MetricsRegistry.RecordWriteBehindError();
                }
                MetricsRegistry.RecordWriteBehindBatch(entries.Length, batchTimer.Elapsed, maxLagSeconds, queueLength);
            }
            catch (RedisException ex)
            {
                _logger.Warn("write-behind read failed", new { Error = ex.Message });
                MetricsRegistry.RecordWriteBehindError();
                await Task.Delay(TimeSpan.FromSeconds(1), stoppingToken).ConfigureAwait(false);
            }
        }
    }

    private async Task ProcessEntryAsync(IDatabase db, StreamEntry entry, CancellationToken cancellationToken)
    {
        try
        {
            var payload = entry["event"];
            if (payload.IsNullOrEmpty)
            {
                await AckAsync(db, entry.Id).ConfigureAwait(false);
                return;
            }
            var writeEvent = JsonSerializer.Deserialize<WriteEvent>(payload!, _jsonOptions);
            if (writeEvent is null || (writeEvent.Source is not null && writeEvent.Source != _source))
            {
                await AckAsync(db, entry.Id).ConfigureAwait(false);
                return;
            }
            await ApplyEventAsync(writeEvent, cancellationToken).ConfigureAwait(false);
            await AckAsync(db, entry.Id).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            _logger.Warn("write-behind apply failed", new { Error = ex.Message });
            MetricsRegistry.RecordWriteBehindError();
        }
    }

    private async Task ApplyEventAsync(WriteEvent writeEvent, CancellationToken cancellationToken)
    {
        switch (writeEvent.Type)
        {
            case WriteEventType.Create when writeEvent.Payload is not null:
                await _store.CreateProductAsync(writeEvent.Payload, cancellationToken).ConfigureAwait(false);
                break;
            case WriteEventType.Update when writeEvent.Payload is not null:
                await _store.UpdateProductAsync(writeEvent.Id, writeEvent.Payload, cancellationToken).ConfigureAwait(false);
                break;
            case WriteEventType.Delete:
                await _store.DeleteProductAsync(writeEvent.Id, cancellationToken).ConfigureAwait(false);
                break;
        }
    }

    private async Task AckAsync(IDatabase db, RedisValue id)
    {
        try
        {
            await db.StreamAcknowledgeAsync(_streamName, _groupName, id).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            _logger.Warn("failed to ack write-behind message", new { Error = ex.Message });
            MetricsRegistry.RecordWriteBehindError();
        }
    }

    private async Task EnsureGroupAsync(CancellationToken cancellationToken)
    {
        try
        {
            var db = _connection.GetDatabase();
            await db.StreamCreateConsumerGroupAsync(_streamName, _groupName, "$", true).ConfigureAwait(false);
        }
        catch (RedisServerException ex) when (ex.Message.Contains("BUSYGROUP", StringComparison.OrdinalIgnoreCase))
        {
            // group already exists
        }
    }

    private static double CalculateLagSeconds(long nowMs, RedisValue id)
    {
        if (id.IsNullOrEmpty)
        {
            return 0;
        }
        var raw = id.ToString();
        var dashIndex = raw.IndexOf('-');
        var timestampSegment = dashIndex >= 0 ? raw[..dashIndex] : raw;
        if (!long.TryParse(timestampSegment, out var producedAt))
        {
            return 0;
        }
        var lagMs = Math.Max(0, nowMs - producedAt);
        return lagMs / 1000d;
    }
}

File: node-service/internal/application/workers/writeBehindWorker.js
const Product = require('../../domain/product');
const metrics = require('../../observability/metrics');

function delay(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

class WriteBehindWorker {
  constructor({ client, streamName, group, consumer, batchSize, blockMs, store, logger, source }) {
    this.client = client;
    this.streamName = streamName;
    this.group = group;
    this.consumer = consumer;
    this.batchSize = batchSize;
    this.blockMs = blockMs;
    this.store = store;
    this.logger = logger;
    this.source = source;
    this.running = false;
    this.loopPromise = null;
  }

  async start() {
    if (this.running || !this.client) {
      return;
    }
    await this.ensureGroup();
    this.running = true;
    this.loopPromise = this.loop();
  }

  async stop() {
    this.running = false;
    if (this.loopPromise) {
      await this.loopPromise.catch(() => {});
    }
  }

  async ensureGroup() {
    try {
      await this.client.xGroupCreate(this.streamName, this.group, '0', { MKSTREAM: true });
    } catch (err) {
      if (!err?.message?.includes('BUSYGROUP')) {
        this.logger?.warn?.('failed to create consumer group', { error: err.message });
      }
    }
  }

  async loop() {
    while (this.running) {
      try {
        const response = await this.client.xReadGroup(
          this.group,
          this.consumer,
          { key: this.streamName, id: '>' },
          { COUNT: this.batchSize, BLOCK: this.blockMs }
        );
        if (!response) {
          continue;
        }
        await this.processStreams(response);
      } catch (err) {
        if (!this.running) {
          return;
        }
        this.logger?.warn?.('write-behind read failed', { error: err.message });
        metrics.recordWriteBehindError();
        await delay(1000);
      }
    }
  }

  async processStreams(streams) {
    const batchStart = process.hrtime.bigint();
    const now = Date.now();
    let processed = 0;
    let maxLagSeconds = 0;
    for (const stream of streams) {
      for (const message of stream.messages) {
        processed += 1;
        const lagSeconds = this.calculateLagSeconds(now, message.id);
        if (lagSeconds > maxLagSeconds) {
          maxLagSeconds = lagSeconds;
        }
        await this.handleMessage(message);
      }
    }
    if (processed === 0) {
      return;
    }
    let queueLength = -1;
    try {
      queueLength = await this.client.xLen(this.streamName);
    } catch (err) {
      this.logger?.warn?.('failed to read queue length', { error: err.message });
      metrics.recordWriteBehindError();
    }
    const durationSeconds = Number(process.hrtime.bigint() - batchStart) / 1e9;
    metrics.recordWriteBehindBatch({
      size: processed,
      durationSeconds,
      lagSeconds: maxLagSeconds,
      queueLength
    });
  }

  async handleMessage(message) {
    try {
      const envelope = message.message || {};
      const rawEvent = envelope.event;
      if (!rawEvent) {
        await this.ack(message.id);
        return;
      }
      const event = JSON.parse(rawEvent);
      if (event.source && event.source !== this.source) {
        await this.ack(message.id);
        return;
      }
      await this.applyEvent(event);
      await this.ack(message.id);
    } catch (err) {
      this.logger?.warn?.('write-behind apply failed', { error: err.message });
      metrics.recordWriteBehindError();
    }
  }

  async ack(id) {
    try {
      await this.client.xAck(this.streamName, this.group, id);
    } catch (err) {
      this.logger?.warn?.('failed to ack message', { id, error: err.message });
      metrics.recordWriteBehindError();
    }
  }

  async applyEvent(event) {
    switch (event.type) {
      case 'create':
        await this.applyCreate(event);
        break;
      case 'update':
        await this.applyUpdate(event);
        break;
      case 'delete':
        await this.applyDelete(event);
        break;
      default:
        this.logger?.warn?.('unknown write event type', { type: event.type });
    }
  }

  hydrateProduct(payload) {
    if (!payload) {
      return null;
    }
    const hydrated = {
      ...payload,
      createdAt: payload.createdAt ? new Date(payload.createdAt) : undefined,
      updatedAt: payload.updatedAt ? new Date(payload.updatedAt) : undefined
    };
    return new Product(hydrated);
  }

  async applyCreate(event) {
    const product = this.hydrateProduct(event.payload);
    if (!product) {
      throw new Error('missing payload for create event');
    }
    await this.store.createProduct(product);
  }

  async applyUpdate(event) {
    const product = this.hydrateProduct(event.payload);
    if (!product) {
      throw new Error('missing payload for update event');
    }
    await this.store.updateProduct(event.id, product);
  }

  async applyDelete(event) {
    await this.store.deleteProduct(event.id);
  }

  calculateLagSeconds(nowMs, id) {
    if (!id) {
      return 0;
    }
    const [timestamp] = String(id).split('-');
    const producedAt = Number(timestamp);
    if (!Number.isFinite(producedAt)) {
      return 0;
    }
    const lagMs = Math.max(0, nowMs - producedAt);
    return lagMs / 1000;
  }
}

module.exports = WriteBehindWorker;
